stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -541,8 +558,88 @@ impl NakamotoBlockProposal {
             builder.load_tenure_info(chainstate, &burn_dbconn, tenure_cause)?;
         let mut tenure_tx = builder.tenure_begin(&burn_dbconn, &mut miner_tenure_info)?;
 
+        let mut replay_txs_maybe: Option<VecDeque<StacksTransaction>> =
+            self.replay_txs.clone().map(|txs| txs.into());
+
         for (i, tx) in self.block.txs.iter().enumerate() {
             let tx_len = tx.tx_len();
+
+            // If a list of replay transactions is set, this transaction must be the next
+            // mineable transaction from this list.
+            if let Some(ref mut replay_txs) = replay_txs_maybe {
+                loop {
+                    let Some(replay_tx) = replay_txs.pop_front() else {
+                        // During transaction replay, we expect that the block only
+                        // contains transactions from the replay set. Thus, if we're here,
+                        // the block contains a transaction that is not in the replay set,
+                        // and we should reject the block.
+                        return Err(BlockValidateRejectReason {
+                            reason_code: ValidateRejectCode::InvalidTransactionReplay,
+                            reason: "Transaction is not in the replay set".into(),
+                        });
+                    };
+                    if replay_tx.txid() == tx.txid() {
+                        break;
+                    }
+
+                    // The included tx doesn't match the next tx in the
+                    // replay set. Check to see if the tx is skipped because
+                    // it was unmineable.
+                    let tx_result = builder.try_mine_tx_with_len(
+                        &mut tenure_tx,
```

Comment:
I just realized that re-using `tenure_tx` here is "incorrect", at least because it's causing us to "double count" tx costs. We need to rollback any cost changes that occur here

Reply:
Sorry after thinking more about this, I'm actually incorrect with this assumption. If the tx succeeds (thus adding to cost), then we error.

Reply:
👍 this is what I was thinking, but then looking at the code, I couldn't find the path during normal block building where the cost of an unmineable transaction is removed, so I wasn't totally sure what was happening without digging more.